src/couch/src/couch_file.erl

```diff
@@ -215,12 +215,26 @@ truncate(Fd, Pos) ->
 sync(Filepath) when is_list(Filepath) ->
     case file:open(Filepath, [append, raw]) of
         {ok, Fd} ->
-            try ok = file:sync(Fd) after ok = file:close(Fd) end;
+            try
+                case file:sync(Fd) of
+                    ok ->
+                        ok;
+                    Else ->
+                        erlang:error({fsync_error, Else})
```

Comment:
Was going to complain about the "surprise" of `erlang:error` here, but I see we use it in a couple of other places, and this is serious enough that it's the only thing I can think of that makes sense.

`erlang:throw` might get caught higher up (I didn't look carefully) and it's good enough for our `couch_file` to die and restart.

If the disk subsystem is really broken, we'll get throttled on trying to restart `couch_file`.

At what point will we fully attempt to restart?

Reply:
What do you mean by "fully restart"? In terms of the database and outstanding writes and re-applying writes that'd be at the first error since couch_file exiting will take down couch_db_updater and friends.

In terms of restarting the entire `beam.smp` process I'd think it'd depend on the error itself. Normally when EIO happens and the filesystem panics into read-only we tend to spew a lot of errors from various places which ends up causing supervision to restart the process eventually. In terms of ENOSPC or similar I think we'd never completely restart unless we happen to bubble up through the auth cache trying to write users or being unable to write docs into the dbs database or some other system related write activity.

Reply:
Also I should say that unlike Postgres we don't have to do a complete restart since our append-only design basically means any successful couch_file open on a shard is basically equivalent to their "restart from valid WAL segment" or whatever they call it.

Or rather, the recovery mode thing they discussed at length we basically perform every time we open a shard since we're seeking back to the last valid header and skipping over any partially written data.

Reply:
Right, my concern is if a disk/volume entirely goes away, and _every single `couch_file`_ starts EIOing out, the entire VM will probably tailspin spewing lots of 500s. I guess this is closer to the read-only panic mode. (I don't want to get into what happens if an NFS-mounted volume goes away, that usually just results in the entire process/kernel hanging on file I/O.)

I think what you're saying is that the supervision tree might try and completely restart in this case, yes? Or does it actually exit, leaving it to `runit`/`systemd`/whatever?

Reply:
Right, if enough of the supervision tree dies often enough eventually `beam.smp` will exit and then rely on a service supervisor to restart it. But that requires taking an entire application down so with IO errors that basically means some sort of CouchDB "system" level activity that ends up taking down an supervised process fast enough.

Normal databases hitting errors will just add churn through couch_server and spew 500s for IO activity hitting IO issues.